src/Campaigns/Blocks/CampaignDonationsBlock/render.php

```diff
@@ -0,0 +1,251 @@
+    class CampaignDonationsBlockRenderer
+    {
+        /**
+         * @var Campaign $campaign
+         */
+        private $campaign;
+
+        /**
+         * @var array $attributes
+         */
+        private $attributes;
+
+        /**
+         * @unreleased
+         */
+        public function __construct($attributes)
+        {
+            $this->attributes = $attributes;
+            $this->campaign = $this->fetchCampaign();
+        }
+
+        /**
+         * @unreleased
+         */
+        private function fetchCampaign(): ?Campaign
+        {
+            if ( ! isset($this->attributes['campaignId'])) {
+                return null;
+            }
+
+            return give(CampaignRepository::class)->getById($this->attributes['campaignId']);
+        }
+
+        /**
+         * @unreleased
+         */
+        public function render(): void
+        {
+            if ( ! $this->campaign) {
+                return;
+            }
+
+            $query = $this->buildDonationsQuery();
+            $donations = $this->formatDonationsData($query->getAll());
+            $this->renderBlockHtml($donations);
+        }
+
+        /**
+         * @unreleased
+         */
+        private function buildDonationsQuery(): CampaignDonationQuery
+        {
+            $sortBy = $this->attributes['sortBy'] ?? 'top-donations';
+            $query = (new CampaignDonationQuery($this->campaign))
+                ->select(
+                    'donation.ID as id',
+                    'donorIdMeta.meta_value as donorId',
+                    'amountMeta.meta_value as amount',
+                    'donation.post_date as date',
+                    'donors.name as donorName'
+                )
+                ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorIdMeta')
+                ->joinDonationMeta(DonationMetaKeys::AMOUNT, 'amountMeta')
+                ->leftJoin('give_donors', 'donorIdMeta.meta_value', 'donors.id', 'donors')
+                ->orderBy($sortBy === 'top-donations' ? 'amount' : 'donation.ID', 'DESC')
+                ->limit($this->attributes['donationsPerPage'] ?? 5);
+
+            if ( ! $this->attributes['showAnonymous']) {
+                $query->joinDonationMeta(DonationMetaKeys::ANONYMOUS, 'anonymousMeta')
+                    ->where('anonymousMeta.meta_value', '0');
+            }
+
+            return $query;
+        }
+
+        /**
+         * @unreleased
+         */
+        private function formatDonationsData(array $donations): array
+        {
+            return array_map(static function ($entry) {
+                $entry->date = human_time_diff(strtotime($entry->date));
+                $entry->amount = Money::fromDecimal($entry->amount, give_get_currency());
+
+                return $entry;
+            }, $donations);
+        }
+
+        /**
+         * @unreleased
+         */
+        private function renderBlockHtml($donations): void
```

Comment:
Oh man, I'm having a hard time reading this 😅
Why don't you move this into a separate file and then use the `View` class to load and render the file?

Reply:
This is a very WP way of doing these things and I have to say that I'm not a big fan 😁

Reply:
Really? I even tried to make it more object oriented, moving away from WP way of doing it.

Reply:
Let's discuss this more.

Reply:
Jon would like to see the template as being a React app.